blog/_src/posts/2018-12-02-java-transient.md

```diff
@@ -0,0 +1,361 @@
+In `average`, these assertions check that: (1) the input is a list,
+ (2) the output is a `float`, (3) and the names `sum` `len` and
+ `ValueError` point to callable values.
+That's all; the assertions **do not check** that `nums` contains only floating-point
+ numbers.
+
+> The assertions also do not check that the function bound to `sum` is defined
+> for a single argument, which is arguably a bug.
+> Scaling a model to an implementation is always challenging.
+
+If `nums` contains something other than floating point numbers, then the
+ call to `average` may cause `sum` to raise an exception or it may silently
+ compute a nonsense result.
+The behavior depends on the implementation of `sum` in the same way that
+ the behavior of a TypeScript function depends on any JavaScript functions
+ that it invokes.
+
+Reticulated does not erase types, nor does it fully enforce types.
+Every type in a Reticulated module translates to its
+ top-level type constructor `C(T)`, e.g.:
+
+```
+  C(Float)                = Float
+  C(List(Float))          = List
+  C(List(Float) -> Float) = ->
+```
+
+Consequently, Reticulated has a slightly stronger run-time guarantee than Python:
+
+- if `e` is an expression with static type `T` that evaluates to a value `v`,
+- then `v` is guaranteed to have a top-level shape that matches the `C(T)`
+  constructor.
+
+
+## Java migratory typing
+
+Java 1.5.0 added [generic types](https://www.jcp.org/en/jsr/detail?id=14)
+ to the Java 1.4.0 type system.
+The benefit of generics is that a programmer can:
+ write one class definition,
+ use the definition in a few different contexts,
+ and receive specific feedback from the type checker in each context.
+
+### Review: generic types
+
+Suppose we want to write a `Box` class that holds some kind of value;
+ the value could be an `Integer` or a `String` or anything else.
+Here is a pre-generics definition:
+
+```
+class Box {
+  private Object val;
+
+  public Box(Object val) { this.set(val); }
+
+  public void set(Object val) { this.val = val; }
+
+  public Object get() { return this.val; }
+}
+```
+
+With this definition is it possible to make boxes that hold different types
+ of values:
+
+```
+// good!
+Box iBox = new Box(new Integer(4));
+Box sBox = new Box(new String("X"));
+```
+
+but it is also possible to "change the type" of the contents of a `Box`:
+
+```
+// maybe bad!
+iBox.set(new String("not a number"));
+```
+
+and some calls to `get` must be followed by a type cast:
+
+```
+// annoying!
+((String) sBox.get()).charAt(0);
+```
+
+**With generics**, we can give a name (e.g. `ValType`) to "the type of the value inside a box":
+
+```
+class GBox<ValType> {
+  private ValType val;
+
+  public GBox(ValType val) { this.set(val); }
+
+  public void set(ValType val) { this.val = val; }
+
+  public ValType get() { return this.val; }
+}
+```
+
+and now we can tell the type checker to check different boxes differently
+ (to opt-out, declare the type as `GBox<Object>`):
```

Comment:
There's also the weird and exists-only-for-backwards-compatibility `GBox` syntax, which essentially gives you `GBox<Object>`, just with much more complaining from the typechecker along the way.